### PR TITLE
Phase out old witness handling in favour of shardtree

### PR DESCRIPTION
**Pull request by AloeareV**

### CODE REVIEW

**Comment by AloeareV** [comment]
This is functionally complete, I have intentionally left it as draft because there's at least one test I want to add to feel confident that it's always doing the correct thing. Also, there's a bunch of noisy extra debugs that I don't want to land in dev

File: zingolib/Cargo.toml [hunk 1]
```diff
@@ -52,20 +52,23 @@ group = "0.13"
...
 subtle = "2.4.1"
-incrementalmerkletree = { workspace = true }
+incrementalmerkletree = { workspace = true, features = ["test-dependencies"] }
 zcash_address = { workspace = true }
 zcash_client_backend = { workspace = true }
 zcash_encoding = { workspace = true }
 zcash_note_encryption = { workspace = true }
 zcash_primitives = { workspace = true }
 zcash_proofs = { workspace = true }
+zcash_client_sqlite = {workspace = true }
```
Comment (zancas) [inline zingolib/Cargo.toml:62]: Do we still need this?

Comment (AloeareV) [reply zingolib/Cargo.toml:62]: Turns out we do! We're using their helpers for `read_shard` and `write_shard`

**Comment by AloeareV** [comment]
> This is functionally complete, I have intentionally left it as draft because there's at least one test I want to add to feel confident that it's always doing the correct thing. Also, there's a bunch of noisy extra debugs that I don't want to land in dev

Also, we need to not bother with witnesses for viewkey wallets

File: zingolib/Cargo.toml (hunk 1 above)
Comment (zancas) [reply zingolib/Cargo.toml:62]: Should we request that those helpers move?

File: zingolib/src/blaze/block_witness_data.rs [hunk 2]
```diff
@@ -645,157 +644,6 @@ impl BlockAndWitnessData {
         )
         .await
     }
-
```
Comment (zancas) [inline zingolib/src/blaze/block_witness_data.rs:648]: I won't be reading this removed code.

File: zingolib/src/blaze/fixed_size_buffer.rs [hunk 3]
```diff
@@ -1,95 +0,0 @@
-use std::usize;
```
Comment (zancas) [inline zingolib/src/blaze/fixed_size_buffer.rs:1]: Not reviewing in detail.

File: zingolib/src/blaze/update_notes.rs [hunk 4]
```diff
@@ -207,23 +119,31 @@ impl UpdateNotes {
...
+                            .add_new_spent(
+                                spent_transaction_id,
+                                spent_at_height,
+                                false,
+                                ts,
+                                nf,
+                                value,
+                                transaction_id,
+                            )
+                            .await;
```
Comment (zancas) [inline zingolib/src/blaze/update_notes.rs:146]: Why is this method *now* async?

File: zingolib/src/wallet/transactions.rs [hunk 5]
```diff
@@ -4,45 +4,41 @@ use std::{
 };
 
 use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
-use incrementalmerkletree::witness::IncrementalWitness;
+use incrementalmerkletree::Position;
```
Comment (zancas) [inline zingolib/src/wallet/transactions.rs:7]: A `Position` is where a leaf is..  so it's a scalar that indexes into an ordered structure.

File: zingolib/src/wallet/transactions.rs [hunk 6]
```diff
@@ -4,45 +4,41 @@ use std::{
...
 use super::{
     data::{
-        OutgoingTxData, PoolNullifier, ReceivedOrchardNoteAndMetadata,
-        ReceivedSaplingNoteAndMetadata, ReceivedTransparentOutput, TransactionMetadata,
-        WitnessCache,
-    },
-    keys::unified::WalletCapability,
-    traits::{
-        Bundle, DomainWalletExt, FromBytes, Nullifier, ReceivedNoteAndMetadata, Recipient, Spend,
+        OutgoingTxData, PoolNullifier, ReceivedTransparentOutput, TransactionMetadata, WitnessTrees,
```
Comment (zancas) [inline zingolib/src/wallet/transactions.rs:24]: Note `super` is `wallet`.

File: zingolib/src/wallet/transactions.rs [hunk 7]
```diff
@@ -55,23 +51,28 @@ impl TransactionMetadataSet {
         mut reader: R,
         wallet_capability: &WalletCapability,
     ) -> io::Result<Self> {
-        let txs_tuples = Vector::read(&mut reader, |r| {
+        let mut witness_trees = match wallet_capability.can_spend() {
```
Comment (zancas) [inline zingolib/src/wallet/transactions.rs:54]: Given that we require all variants to be able to spend, should we abstract a new method that's just a bool for `can_spend`.   That's how we're using it.. but with the potential for misunderstanding...

Surely in the (currently impossible, in Zingo) case of spend capabilities from some pools and viewing from others, the holder doesn't intend to not be able to spend.

It seems to me that `can_spend` is at least an inadequate name.  How about we rename it to:

`can_spend_from_any_pool`?

Comment (zancas) [reply zingolib/src/wallet/transactions.rs:54]: Or rather..   I am going to add `can_spend_from_all_pools -> bool`
Then we should call that, when that's what we mean.

Comment (zancas) [reply zingolib/src/wallet/transactions.rs:54]: In all instances of in-library calls, what we we're actually checking for is spend_from_all_pools.

File: zingolib/src/wallet/data.rs [hunk 8]
```diff
@@ -36,6 +42,199 @@ pub enum PoolNullifier {
...
+        Self {
+            witness_tree_sapling: shardtree::ShardTree::new(MemoryShardStore::empty(), MAX_REORG),
+            witness_tree_orchard: shardtree::ShardTree::new(MemoryShardStore::empty(), MAX_REORG),
+        }
+    }
+}
+
+impl WitnessTrees {
+    pub(crate) fn add_checkpoint(&mut self, height: BlockHeight) {
+        self.witness_tree_sapling.checkpoint(height).unwrap();
```
Comment (zancas) [inline zingolib/src/wallet/data.rs:154]: I don't understand the details of checkpoint..  I think I'd need to write it out.

File: zingolib/src/lightclient.rs [hunk 9]
```diff
@@ -1355,6 +1376,25 @@ impl LightClient {
         // The top of the wallet
         let last_synced_height = self.wallet.last_synced_height().await;
 
+        self.ensure_witness_tree_not_above_wallet_blocks().await;
```
Comment (zancas) [inline zingolib/src/lightclient.rs:1379]: Let's a add comment explaining the scenario that could lead to the witness tree being ahead of the wallet blocks.

File: zingolib/src/lightclient.rs [hunk 10]
```diff
@@ -245,6 +245,27 @@ impl LightClient {
         };
         LightClient::read_wallet_from_buffer(config, BufReader::new(File::open(wallet_path)?))
     }
+
```
Comment (zancas) [inline zingolib/src/lightclient.rs:248]: Let's add a comment here that explains how this fn ensures that the witness tree is..  uh...   bound to the wallet last scanned height witness tree?

File: zingolib/src/lightclient.rs [hunk 11]
```diff
@@ -1522,11 +1562,11 @@ impl LightClient {
                 .await;
 
         // Do Trial decryptions of all the outputs, and pass on the successful ones to the update_notes processor
-        let trial_decryptions_processor = TrialDecryptions::new(
+        let trial_decryptions_processor = Arc::new(TrialDecryptions::new(
```
Comment (zancas) [inline zingolib/src/lightclient.rs:1565]: Add a comment explaining why the tdp now must be in an `Arc`

File: zingolib/src/wallet/traits.rs [hunk 12]
```diff
@@ -917,29 +995,24 @@ where
...
         spend_key: Option<&D::SpendingKey>,
     ) -> Option<Self> {
         // Include only notes that haven't been spent, or haven't been included in an unconfirmed spend yet.
         if note_and_metadata.spent().is_none()
-            && note_and_metadata.unconfirmed_spent().is_none()
+            && note_and_metadata.pending_spent().is_none()
             && spend_key.is_some()
-            && note_and_metadata.witnesses().len() >= (anchor_offset + 1)
+            && !note_and_metadata.pending_receipt()
+        //TODO: Account for lack of this line
```
Comment (zancas) [inline zingolib/src/wallet/traits.rs:1005]: Do we still need to handle this case?

File: zingolib/src/wallet/transactions.rs (hunk 7 above)
Comment (AloeareV) [reply zingolib/src/wallet/transactions.rs]: I see that you've added this. I like it!

File: zingolib/src/lightclient.rs (hunk 11 above)
Comment (AloeareV) [reply zingolib/src/lightclient.rs]: Turns out there's no longer a good reason!

File: zingolib/src/lightclient.rs (hunk 10 above)
Comment (AloeareV) [reply zingolib/src/lightclient.rs:248]: I don't understand what you mean

File: zingolib/src/wallet/data.rs (hunk 8 above)
Comment (AloeareV) [reply zingolib/src/wallet/data.rs:154]: I think we delved into this sufficiently via screenshare

File: zingolib/src/lightclient.rs (hunk 9 above)
Comment (AloeareV) [reply zingolib/src/lightclient.rs:1375]: Done

**Review by zancas — APPROVED** [review]
This is a major upgrade!

**Pull request merged**